### PR TITLE
MAXSwerve drive branch

### LINKED ISSUE #1: Manual Driving
## Description: Follow the task and import the swerve needed to get the code running (autonomous code is not needed for this issue)

# Steps needed to complete this task : 

1. - [ ] Set up a constants class - This requires #2 to be complete 

2. - [ ] Create/import Swerve Module class 

3. - [ ] Create/import Set up Swerve Driving class 

4. - [ ] Create controls/commands to move the robot, place it in the Robot container Don't forget to:

- Create a branch off of main
  - Name the branch related to the feature -- e.g, drive code might be named `drive`
- Commit & Push to GitHub often -- You want to capture major milestones 
- Create a pull request when complete. 
  - The description should include a brief description and say `This resolves #TICKETNUM.`
  - The build should pass -- make sure you get a ✅ in the pull request. If ❌, you need to ensure your code compiles.
    - If the `Run Format Check` fails ❌, you need to run `.\gradlew spotlessJavaApply` in Powershell/CMD (cntl ~ to open the console in VSCode)  
- The pull request will need to be checked by a peer/ or programming lead, and then a mentor 

## Acceptance Criteria:
- Robot is able to drive in straight lines in any direction 
- Robot is able to rotate left and right 
- Robot is able to move while spinning
- Robot is able to do a figure 8 while turning

### PR DESCRIPTION
**Pull request by QuantumOtterDev**
Created drive/module class to manually drive the robot for tele-op

Code was tested on robot Saturday 1/15 - Robot moved how it should; no errors were thrown

### CODE REVIEW

File: src/main/java/frc/lib/swerve/MAXSwerve.java [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+    m_maxSpeed = maxSpeed;
+    m_odometry = new SwerveDriveOdometry(kinematics, m_gyro.getRotation2d(), modulePositions);
+  }
+
+  private ChassisSpeeds m_chassisSpeed = new ChassisSpeeds();
+
+  @Override
+  public void initSendable(SendableBuilder builder) {
+    super.initSendable(builder);
+    // builder.addDoubleProperty("Velocity vx", () -> m_chassisSpeed.vxMetersPerSecond, null);
```
Comment (nbgray) [inline src/main/java/frc/lib/swerve/MAXSwerve.java]: Please remove commented out code, or add a TODO indicating what you want to do with it.

File: src/main/java/frc/lib/swerve/MAXSwerve.java [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+   */
+  public void drive(double xSpeed, double ySpeed, double rot, boolean fieldRelative) {
+    // If nothing is commanded, hold the same position - don't know if the rev modules need this
+    // yet?
+    if (xSpeed == 0.0 && ySpeed == 0.0 && rot == 0.0) {
+      holdAllModulesRotation();
+      return;
+    }
+    // Adjust input based on max speed
+    xSpeed *= m_maxSpeed;
```
Comment (nbgray) [inline src/main/java/frc/lib/swerve/MAXSwerve.java:196]: Do you want to multiple xSpeed by m_maxSpeed?  Or limit it to it with xSpeed = min(xSpeed, m_maxSpeed)?
Same for ySpeed.

File: src/main/java/frc/lib/swerve/MAXSwerve.java [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+    SwerveDriveKinematics.desaturateWheelSpeeds(swerveModuleStates, m_maxSpeed);
+    m_frontLeft.setDesiredState(swerveModuleStates[0]);
+    m_frontRight.setDesiredState(swerveModuleStates[1]);
+    m_backLeft.setDesiredState(swerveModuleStates[2]);
+    m_backRight.setDesiredState(swerveModuleStates[3]);
+  }
+
+  private void holdModuleRotation(MAXModule m) {
+    var state = m.getDesiredState();
+    state.speedMetersPerSecond = 0;
```
Comment (nbgray) [inline src/main/java/frc/lib/swerve/MAXSwerve.java]: state/speedMetersPerSecond is a double, so it should be set to 0.0

File: src/main/java/frc/robot/Constants.java [hunk 4]
```diff
@@ -12,4 +20,163 @@
...
+    public static final boolean kTurningEncoderInverted = true;
+
+    // Calculations required for driving motor conversion factors and feed forward
+    public static final double kDrivingMotorFreeSpeedRps = NeoMotorConstants.kFreeSpeedRpm / 60;
+    public static final double kWheelDiameterMeters = 0.0762;
+    public static final double kWheelCircumferenceMeters = kWheelDiameterMeters * Math.PI;
+    // 45 teeth on the wheel's bevel gear, 22 teeth on the first-stage spur gear, 15 teeth on the
+    // bevel pinion
+    public static final double kDrivingMotorReduction =
+        (45.0 * 22) / (kDrivingMotorPinionTeeth * 15);
```
Comment (nbgray) [inline src/main/java/frc/robot/Constants.java]: 22 should be 22.0, 15 should be 15.0

File: src/main/java/frc/robot/Constants.java [hunk 5]
```diff
@@ -12,4 +20,163 @@
...
+    // This changes the drive speed of the module (a pinion gear with more teeth will result in a
+    // robot that drives faster).
+    public static final int kDrivingMotorPinionTeeth = 14;
+
+    // Invert the turning encoder, since the output shaft rotates in the opposite direction of
+    // the steering motor in the MAXSwerve Module.
+    public static final boolean kTurningEncoderInverted = true;
+
+    // Calculations required for driving motor conversion factors and feed forward
+    public static final double kDrivingMotorFreeSpeedRps = NeoMotorConstants.kFreeSpeedRpm / 60;
```
Comment (nbgray) [inline src/main/java/frc/robot/Constants.java]: 60 should be 60.0

File: src/main/java/frc/robot/Constants.java [hunk 6]
```diff
@@ -12,4 +20,163 @@
...
+    public static final double kDrivingEncoderPositionFactor =
+        (kWheelDiameterMeters * Math.PI) / kDrivingMotorReduction; // meters
+    public static final double kDrivingEncoderVelocityFactor =
+        ((kWheelDiameterMeters * Math.PI) / kDrivingMotorReduction) / 60.0; // meters per second
+
+    public static final double kTurningEncoderPositionFactor = (2 * Math.PI); // radians
+    public static final double kTurningEncoderVelocityFactor =
+        (2 * Math.PI) / 60.0; // radians per second
+
+    public static final double kTurningEncoderPositionPIDMinInput = 0; // radians
```
Comment (nbgray) [inline src/main/java/frc/robot/Constants.java]: 0 should be 0.0

File: src/main/java/frc/robot/Constants.java [hunk 7]
```diff
@@ -12,4 +20,163 @@
...
+    public static final double kTurningEncoderPositionPIDMaxInput =
+        kTurningEncoderPositionFactor; // radians
+
```
Comment (nbgray) [inline src/main/java/frc/robot/Constants.java:139]: Ok just in general all doubles should be set as x.x using a decimal.  Floats should be x.xf.  This is the pattern you should follow for all places you set a double, from declaration to method calls.

File: src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        new SwerveModulePosition[] {
+          frontLeft.getPosition(),
+          frontRight.getPosition(),
+          backLeft.getPosition(),
+          backRight.getPosition()
+        },
+        gyro,
+        Constants.SwerveDriveConstants.kMaxSpeedMetersPerSecond);
+  }
+
```
Comment (nbgray) [inline src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:50]: Commented out code should be removed.

**Review by nbgray — CHANGES_REQUESTED** [review]
Your build fails online.  It looks like an issue with the pathplanner.  This must be resolved before the changes can be merged.

File: src/main/java/frc/lib/swerve/MAXSwerve.java (hunk 2 above)
Comment (QuantumOtterDev) [reply src/main/java/frc/lib/swerve/MAXSwerve.java:196]: you want to multiply it

**Review by nbgray — CHANGES_REQUESTED** [review]
It looks like https://3015rangerrobotics.github.io/pathplannerlib/repo/ is down.  Until that comes back up any builds that haven't downloaded the maven repo will be unable to, and the build will fail for them.  It works for you because you already have it downloaded.  See if you can find a backup location for the repo, or contact 3015 to let them know their repo page is down.

**Review by nbgray — APPROVED** [review]
Changes look good.  pathplanner issue is resolved.

**Pull request merged**